lib/engines/csvEngine.js

```diff
@@ -0,0 +1,163 @@
+var _ = require("lodash");
+var Bromise = require("bluebird");
+var converter = require("json-2-csv");
+var mqtt = require("mqtt");
+var fs = require('fs');
+
+var MQTT_PORT = "1883";
+
+/***
+Config:
+   "generation": {
+        "mode": "csvEngine",
+        "csvEngine": {
+            "fileRepository": ".",
+            "csvOptions": "",
+            "timestamp": 1,
+            "value": 2
+            }
+        }
+***/
+
+function CSVEngine(simulator) {
+
+    this.config = simulator.simulation.generation.csvEngine;
+    this.simulator = simulator;
+    this.mqttServer = "tcp://" + simulator.configuration.server;
+    this.dataGenerators = {};
+}
+
+CSVEngine.prototype.generateValues = function(dataDescriptor, system) {
+    var self = this;
+    var result = {};
+
+    var filename = this.config.fileRepository + "/" + system.name + ".csv";
+    console.log("    Reading data [" + filename + "]");
+    var config = this.config;
```

Comment:
No need to do this here starting from NodeJS 4.x we can use what is called **arrow functions** which keep the original context and allows us to do 

```js
return new Bromise((resolve, reject) => {
   fs.readFile(filename, 'utf-8', (err, data) => {
      // ...
      converter.csv2json(data, (err, json) => {
          // ...
      }, this.config.csvOptions);
   });
});
```

But it would probably be better to split this function in smaller ones. I'll do it.

Reply:
Replace return new Bromise(function(resolve, reject) {
        fs.readFile(filename, 'utf8', function(err, data) {

Removed var config = this.config; not used anymore
update converter.csv2json call and this.config.csvOptions line